crates/telio-starcast/src/utils.rs

```diff
@@ -0,0 +1,317 @@
+use pnet_packet::{
+    ip::IpNextHeaderProtocol,
+    ipv4::{Ipv4Packet, MutableIpv4Packet},
+    ipv6::MutableIpv6Packet,
+    MutablePacket,
+};
+use std::net::IpAddr;
+
+pub trait MutableIpPacket<'a>: Sized + MutablePacket {
+    fn new(buffer: &'a mut [u8]) -> Option<Self>;
+
+    fn get_next_level_protocol(&self) -> IpNextHeaderProtocol;
+    fn get_source(&self) -> IpAddr;
+    fn get_destination(&self) -> IpAddr;
+
+    fn set_source(&mut self, addr: IpAddr);
+    fn set_destination(&mut self, addr: IpAddr);
+    fn set_checksum(&mut self, addr: u16);
+
+    fn to_immutable_ipv4<'p>(&'p self) -> Ipv4Packet<'p>;
+}
+
+impl<'a> MutableIpPacket<'a> for MutableIpv4Packet<'a> {
+    fn new(buffer: &'a mut [u8]) -> Option<Self> {
+        Self::new(buffer)
+    }
+
+    fn get_next_level_protocol(&self) -> IpNextHeaderProtocol {
+        self.get_next_level_protocol()
+    }
+
+    fn get_destination(&self) -> IpAddr {
+        self.get_destination().into()
+    }
+
+    fn get_source(&self) -> IpAddr {
+        self.get_source().into()
+    }
+
+    fn set_source(&mut self, addr: IpAddr) {
+        if let IpAddr::V4(ipv4addr) = addr {
+            self.set_source(ipv4addr);
+        }
+    }
+
+    fn set_destination(&mut self, addr: IpAddr) {
+        if let IpAddr::V4(ipv4addr) = addr {
+            self.set_destination(ipv4addr);
+        }
+    }
+
+    fn set_checksum(&mut self, checksum: u16) {
+        self.set_checksum(checksum);
+    }
+
+    fn to_immutable_ipv4<'p>(&'p self) -> Ipv4Packet<'p> {
+        self.to_immutable()
+    }
+}
+
+impl<'a> MutableIpPacket<'a> for MutableIpv6Packet<'a> {
+    fn new(buffer: &'a mut [u8]) -> Option<Self> {
+        Self::new(buffer)
+    }
+
+    fn get_next_level_protocol(&self) -> IpNextHeaderProtocol {
+        self.get_next_header()
+    }
+
+    fn get_destination(&self) -> IpAddr {
+        self.get_destination().into()
+    }
+
+    fn get_source(&self) -> IpAddr {
+        self.get_source().into()
+    }
+
+    fn set_destination(&mut self, addr: IpAddr) {
+        if let IpAddr::V6(ipv6addr) = addr {
+            self.set_destination(ipv6addr);
+        }
+    }
+
+    fn set_source(&mut self, addr: IpAddr) {
+        if let IpAddr::V6(ipv6addr) = addr {
+            self.set_source(ipv6addr);
+        }
+    }
+
+    fn set_checksum(&mut self, _: u16) {
+        unimplemented!()
```

Comment:
leftover ?

Reply:
Nope, they should not be used. There is no checksum to set in ipv6 packet and to_immutable_ipv4 also does not make sense there. With the proper ut coverage I feel safe leaving them as panicking, but let me know if you want me to change them to no-op + comment

Reply:
So add a comment at least

Reply:
This comment is obsolete in the new implementation. Resolving